lib/livebook/runtime/k8s.ex

```diff
@@ -0,0 +1,566 @@
+
+      {:error, error} ->
+        {:error, "could not create Pod, reason: #{Exception.message(error)}"}
+    end
+  end
+
+  defp get_cluster_data(nil), do: %{node_base: "k8s_runtime", remote_port: 44444}
+
+  defp get_cluster_data(_kubeconfig) do
+    local_port = get_free_port!()
+    %{node_base: "remote_runtime_#{local_port}", remote_port: 44444, local_port: local_port}
+  end
+
+  defp k8s_forward_port(nil, _, _, _, _), do: :ok
+
+  defp k8s_forward_port(
+         kubeconfig,
+         context,
+         %{local_port: local_port, remote_port: remote_port},
+         pod_name,
+         namespace
+       ) do
+    with {:ok, kubectl_path} <- find_kubectl_executable() do
+      ports = "#{local_port}:#{remote_port}"
+
+      # We want the proxy to accept the same protocol that we are
+      # going to use for distribution
+      bind_addr =
+        if Livebook.Utils.proto_dist() == :inet6_tcp do
+          "[::1]"
+        else
+          "127.0.0.1"
+        end
+
+      args =
+        [
+          "port-forward",
+          "--kubeconfig",
+          Path.expand(kubeconfig),
+          "--context",
+          context,
+          "-n",
+          namespace,
+          pod_name,
+          ports,
+          "--address",
+          bind_addr
+        ]
+
+      port =
+        Port.open(
+          {:spawn_executable, kubectl_path},
+          [:binary, :hide, :stderr_to_stdout, args: args, env: []]
+        )
+
+      port_ref = Port.monitor(port)
+
+      result =
+        receive do
+          {^port, {:data, "Forwarding from " <> _}} ->
+            :ok
+
+          {^port, {:data, "Error: " <> error}} ->
+            {:error, "failed to port-forward. Error: #{String.trim(error)}"}
+
+          {:DOWN, ^port_ref, :port, _object, reason} ->
+            {:error, "failed to port-forward. Process terminated, reason: #{inspect(reason)}"}
+        after
+          30_000 ->
+            {:error, "failed to port-forward. Timed out after 30s"}
+        end
+
+      Port.demonitor(port_ref, [:flush])
+
+      result
+    end
+  end
+
+  defp find_kubectl_executable() do
+    if path = System.find_executable("kubectl") do
+      {:ok, path}
+    else
+      {:error, "no kubectl executable found in PATH."}
+    end
+  end
+
+  defp await_pod_ready(req, namespace, pod_name) do
+    with :ok <-
+           Kubereq.wait_until(
+             req,
+             namespace,
+             pod_name,
+             fn
+               :deleted ->
+                 {:error, "The Pod was deleted before it started running."}
+
+               pod ->
+                 # send(caller, {:runtime_connect_info, pid, pod[]})
+                 get_in(pod, [
+                   "status",
+                   "conditions",
+                   Access.filter(&(&1["type"] == "Ready")),
+                   "status"
+                 ]) == ["True"]
+             end,
+             # 24 hours (try "forever" and let peopel stop manually)
+             86_400_000
```

Comment:
I'm having a hard time setting this to a fix number. Sometimes it can take a few minutes for sure. First the scheduler has to find a node that suits the needs. Maybe a new node needs to be provisioned first. Once the pod can be scheduled, the image has to be downloaded. 

Also see the comment on the abort functionality.

Will set to 10 minutes for now.

Reply:
Any reasonable upper bound work for me, can be 30 minutes. I just wouldn't make it hours.